controllers/src/main/java/io/syndesis/controllers/extension/ExtensionDataProvider.java

```diff
@@ -0,0 +1,70 @@
+/**
+ * Copyright (C) 2016 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.syndesis.controllers.extension;
+
+import io.syndesis.dao.manager.DataManager;
+import io.syndesis.filestore.FileStore;
+import io.syndesis.model.extension.Extension;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nonnull;
+import java.io.InputStream;
+import java.util.Set;
+
+/**
+ * Provide information on installed tech extensions.
+ */
+@Component
+@ConditionalOnProperty(name = "features.filestore.enabled")
+public class ExtensionDataProvider {
+
+    private final DataManager dataManager;
+
+    private final FileStore fileStore;
+
+    public ExtensionDataProvider(DataManager dataManager, FileStore fileStore) {
+        this.dataManager = dataManager;
+        this.fileStore = fileStore;
+    }
+
+    public Extension getExtensionMetadata(String extensionId) {
+        String id = getInstalledExtensionId(extensionId);
+        return dataManager.fetch(Extension.class, id);
+    }
+
+    public InputStream getExtensionBinaryFile(String extensionId) {
+        String id = getInstalledExtensionId(extensionId);
+        return fileStore.read("/extensions/" + id);
+    }
+
+    // ==========================================================
+
+    @Nonnull
+    private String getInstalledExtensionId(String extensionId) {
```

Comment:
The name is a little bit misleading as it seems it returns the extensionid by giving an extensionid, maybe something like getInstalledPhysicalId or whatever make it more clear.

Reply:
Right!